validator/client/beacon-api/duties.go

```diff
@@ -5,21 +5,228 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"net/url"
 	"strconv"
 
 	"github.com/pkg/errors"
 	"github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
 	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
 )
 
 type dutiesProvider interface {
 	GetAttesterDuties(ctx context.Context, epoch types.Epoch, validatorIndices []types.ValidatorIndex) ([]*apimiddleware.AttesterDutyJson, error)
+	GetProposerDuties(ctx context.Context, epoch types.Epoch) ([]*apimiddleware.ProposerDutyJson, error)
+	GetSyncDuties(ctx context.Context, epoch types.Epoch, validatorIndices []types.ValidatorIndex) ([]*apimiddleware.SyncCommitteeDuty, error)
+	GetCommittees(ctx context.Context, epoch types.Epoch) ([]*apimiddleware.CommitteeJson, error)
 }
 
 type beaconApiDutiesProvider struct {
 	jsonRestHandler jsonRestHandler
 }
 
+type committeeIndexSlotPair struct {
+	committeeIndex types.CommitteeIndex
+	slot           types.Slot
+}
+
+func (c beaconApiValidatorClient) getDuties(ctx context.Context, in *ethpb.DutiesRequest) (*ethpb.DutiesResponse, error) {
+	multipleValidatorStatus, err := c.multipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{PublicKeys: in.PublicKeys})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get validator status")
+	}
+
+	// Sync committees are an Altair feature
+	fetchSyncDuties := in.Epoch >= params.BeaconConfig().AltairForkEpoch
+
+	currentEpochDuties, err := c.getDutiesForEpoch(ctx, in.Epoch, multipleValidatorStatus, true, fetchSyncDuties)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get duties for current epoch `%d`", in.Epoch)
+	}
+
+	// We don't fetch proposer duties for epoch+1
```

Comment:
Actually we do: https://github.com/prysmaticlabs/prysm/blob/9f44d6e45292315f6103159d2ef88d9c0c900ea0/beacon-chain/rpc/eth/validator/validator.go#L153